src/main/java/moze_intel/projecte/gameObjs/items/rings/TimeWatch.java

```diff
@@ -171,22 +160,101 @@ private void speedUpRandomTicks(Level level, int bonusTicks, AABB bBox) {
-						state.randomTick(serverLevel, pos, serverLevel.random);
+			// optimistically check if the block is already sped up and if so, just increase the bonus ticks
+			if (levelRandomTicks.containsKey(pos.asLong())) {
+				levelRandomTicks.get(pos.asLong()).addTicks(bonusTicks);
+			} else {
+				if (WorldHelper.isBlockLoaded(serverLevel, pos)) {
+					BlockState state = serverLevel.getBlockState(pos);
+					Block block = state.getBlock();
+					if (state.isRandomlyTicking() && !state.is(PETags.Blocks.BLACKLIST_TIME_WATCH)
+						&& !(block instanceof LiquidBlock) // Don't speed non-source fluid blocks - dupe issues
+						&& !(block instanceof BonemealableBlock) && !(block instanceof IPlantable)) {// All plants should be sped using Harvest Goddess
+						pos = pos.immutable();
+						levelRandomTicks.put(pos.asLong(), new SpedUpRandomBlock(serverLevel, pos, state, bonusTicks));
 					}
 				}
 			}
 		}
 	}
 
+	public static void completeTick(Level level, BooleanSupplier haveTime) {
+		long currentTime = System.currentTimeMillis();
+		double maxTPSLost = ProjectEConfig.server.effects.maxTPSLoss.get();
+		double asMS;
+		if (maxTPSLost >= 20 || maxTPSLost < 0) {
+			asMS = 10000;
+		} else {
+			asMS = 1000d / (20 - maxTPSLost) - 50;
+		}
+		// we now need to turn the 2 queues into a schedule that overshoots by the same ratio of ticks done to ticks scheduled for each
+		// we start by combining the 2 queues into a single one
+		// add up the total ticks,
+		// and allocate time accordingly
+
+		int totalTicks = 0;
+		int divisor = -1;
+		List<SpedUpItem> queue = new ArrayList<>();
+		Long2ObjectOpenHashMap<TimeWatch.SpedUpBlockEntity> qBE = queuedBlockEntityTicks.get(level);
+		if (qBE != null) {
+			for (SpedUpItem item : qBE.values()) {
+				queue.add(item);
+				if (divisor == -1 || item.getTicksLeft() < divisor) {
+					divisor = item.getTicksLeft();
+				}
+				totalTicks += item.getTicksLeft();
+			}
+			qBE.clear();
+		}
+		Long2ObjectOpenHashMap<TimeWatch.SpedUpRandomBlock> qRT = queuedRandomTicks.get(level);
+		if (qRT != null) {
+			for (SpedUpItem item : qRT.values()) {
+				queue.add(item);
+				if (divisor == -1 || item.getTicksLeft() < divisor) {
+					divisor = item.getTicksLeft();
+				}
+				totalTicks += item.getTicksLeft();
+			}
+			qRT.clear();
+		}
+		Set<SpedUpItem> toRemove = new HashSet<>();
+
+		int ticksTot = 0;
+
+		while (!queue.isEmpty()) {
+			if (haveTime.getAsBoolean()) currentTime = System.currentTimeMillis();
+			else if (System.currentTimeMillis() - currentTime > asMS) {
+				// we have run out of time, so we will drop the rest of the queue
+//				System.out.println("TimeWatch: Ran out of time, dropping " + (totalTicks - ticksTot) + " extra block ticks");
+				break;
+			}
+
+			// now we run for ticksleft / divisor ticks
+			for (SpedUpItem item : queue) {
+
+				// reasonable range so a single doesn't hog too much
+				int ticks = Mth.clamp(item.getTicksLeft() / divisor, 1, Math.min(totalTicks, 1_000));
```

Comment:
there's probably a better function usable here, 

the idea is to get a ratio so that things that have 100 bonus ticks don't get the same as things with only like 10. but we also don't want this while loop to take too long for a single iteration, since it only checks if it's out of time each loop